qcodes/instrument_drivers/Harvard/Decadac.py

```diff
@@ -1,266 +1,516 @@
-import logging
-from time import sleep
+from time import time
 from functools import partial
-from qcodes.instrument.visa import VisaInstrument
+from qcodes import VisaInstrument, InstrumentChannel, ChannelList, ManualParameter
 from qcodes.utils import validators as vals
 
-log = logging.getLogger(__name__)
+class DACException(Exception):
+    pass
 
+class DacReader(object):
+    @staticmethod
+    def _dac_parse(resp):
+        """
+        Parses responses from the DAC. They should take the form of "<cmd><resp>!"
+        This command returns the value of resp.
+        """
+        resp = resp.strip()
+        if resp[-1] != "!":
+            raise DACException("Unexpected terminator on response: {}. Should end with \"!\"".format(resp))
+        return resp.strip()[1:-1]
 
-class Decadac(VisaInstrument):
-    """
-    The qcodes driver for the Decadac.
-    Each slot on the Deacadac is to be treated as a seperate
-    four-channel instrument.
-
-    Tested with a Decadec firmware revion number 14081 (Decadac 139).
-
-    The message strategy is the following: always keep the queue empty, so
-    that self.visa_handle.ask(XXX) will return the answer to XXX and not
-    some previous event.
+    def _dac_v_to_code(self, volt):
+        """
+        Convert a voltage to the internal dac code (number between 0-65536)
+        based on the minimum/maximum values of a given channel.
+        Midrange is 32768.
+        """
+        frac = (volt - self.min_val) / (self.max_val - self.min_val)
+        val = int(round(frac * 65536))
+        if val >= 65536: # Check limits. For example setting max_val will cause an overflow
+            return 65535
+        if val < 0: # Ensure no negative values
+            return 0
+        return val
+
+    def _dac_code_to_v(self, code):
+        """
+        Convert a voltage to the internal dac code (number between 0-65536)
+        based on the minimum/maximum values of a given channel.
+        Midrange is 32768.
+        """
+        frac = code/65536.0
+        return (frac * (self.max_val - self.min_val)) + self.min_val
 
+    def _set_slot(self):
+        """
+        Set the active DAC slot
+        """
+        resp = self.ask_raw("B{};".format(self._slot))
+        if int(self._dac_parse(resp)) != self._slot:
+            raise DACException("Unexpected return from DAC when setting slot: {}. DAC slot may not have been set.".format(response))
```

Comment:
Yep, this is correct. I will run some more tests on the Decadac driver tomorrow and make these changes.